pkg/cloudprovider/aws/aws.go

```diff
@@ -313,3 +328,96 @@ func (n *NodeGroup) setASGDesiredSize(newSize int64) error {
 	_, err := n.provider.service.SetDesiredCapacity(input)
 	return err
 }
+
+// setASGDesiredSizeOneShot uses the AWS fleet API to acquire all desired
+// capacity in one step and then add it to the existing auto-scaling group.
+func (n *NodeGroup) setASGDesiredSizeOneShot(addCount int64) error {
+	fleet, err := n.provider.ec2_service.CreateFleet(&ec2.CreateFleetInput{
+		Type:                             awsapi.String("instant"),
+		TerminateInstancesWithExpiration: awsapi.Bool(false),
+		OnDemandOptions: &ec2.OnDemandOptionsRequest{
+			MinTargetCapacity:      awsapi.Int64(addCount),
+			SingleAvailabilityZone: awsapi.Bool(true),
```

Comment:
I've just tested adding override options with each of the three AZs and subnets I want the instances to be balanced across, and it still seems like they get created in the one AZ - us-west-2.

Reply:
I removed the `SingleAvailabilityZone`, that was for some early testing and I forgot to remove it. Without that flag and with multiple AZs it should be possible to balance across AZs. I'll do some more testing and update with more details.

Reply:
Dug into this a little more and it looks like AZ balancing isn't yet supported with the `CreateFleet` API. The current use-case I'm solving for by integrating `CreateFleet` support into Escalator is to handle workloads that would benefit from having capacity available immediately without waiting for an ASG to scale-up and are not concerned with AZ balance (for example: an ML training workload that needs p3dn capacity and may be using placement groups to manage network locality).

I think there are two paths forward here. The first would be to merge this as-is and update the docs to state that this is designed to support workloads that don't need AZ balance. I can also follow up with the fleet team to see if AZ balance is on their roadmap and advocate for it. This would give us a path towards updating this code in the future when/if that becomes supported.

The second option is to break the fleet request into multiple parts, one per AZ, and handle balancing within the Escalator code. I think this is risky because there are a lot of edge cases around acquiring the correct amount of capacity at the right time and managing over/under-scaling manually in Escalator and would strongly prefer option one if you're open to it.

WDYT? Do you see different options?

Reply:
Thanks for the detailed explanation, I'm happy with option 1. I definitely agree the second option isn't ideal and could potentially be a bit of a hack.

Could you update the docs in this PR to include the AZ balancing caveat for fleet?  It would be awesome if you could advocate for AZ balancing for fleet as it's something that we could utilise too.

I assume most users that are doing ML training workloads aren't concerned with all of the workloads running in a single AZ?

Reply:
I've updated the docs and will start to push forward AZ balancing with my partner team that owns this API.

As far as the ML training workloads go it's highly desirable that instances be near each other from a networking perspective to cut inter-instance latencies. This patch will also allow users to configure placement groups within their instance profile to further restrict placement for highly latency sensitive workloads.